cmd/agent/app/server.go

```diff
@@ -97,15 +102,50 @@ func (a *Agent) runProxyConnection(o *options.GrpcProxyAgentOptions, stopCh <-ch
 	cs := cc.NewAgentClientSet(stopCh)
 	cs.Serve()
 
-	return nil
+	return cs, nil
 }
 
-func (a *Agent) runHealthServer(o *options.GrpcProxyAgentOptions) error {
+func (a *Agent) runHealthServer(o *options.GrpcProxyAgentOptions, cs agent.ReadinessManager) error {
 	livenessHandler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
 		fmt.Fprintf(w, "ok")
 	})
+
+	checks := []agent.HealthChecker{agent.PingHealthz}
+
+	if o.EnableAgentServerReadiness {
+		checks = append(checks, agent.NewServerConnectedHealthz(cs))
+	}
+
 	readinessHandler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		fmt.Fprintf(w, "ok")
+		var failedChecks []string
```

Comment:
This section is very close to fully generic. It would be nice to organize it somewhere different than cmd/agent to keep it minimal. It would also be good to add some unit test coverage (for example: exercise the verbose codepath).

I'm ok with deferring these suggestions, to unblock making the feature available.

Reply:
IMO, I'd like to get this merged as is.

 The addition of unit tests, while useful would require more thoughts and time and the result would be a very trivial check and not worth the effort and time.